File: lintcode/binary_tree_inorder.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Inorder in ArrayList which contains node values.
    """
    def inorderTraversal(self, root):
        result, node = [], root
        while node:
            if node.left:
                pre = node.left
                while pre.right and pre.right != node:
                    pre = pre.right
                if pre.right:
                    result.append(node.val)
                    node, pre.right = node.right, None
                else:
                    pre.right, node = node, node.left
            else:
                result.append(node.val)
                node = node.right
        return result

File: leetcode/binary_tree_inorder.js
/* Given a binary tree, return the inorder traversal of its nodes' values.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    result = []
    node = root   
    while (node) {
        if (node.left) {
            for (pre = node.left; pre.right && pre.right !== node;)
                pre = pre.right
            if (pre.right) {
                result.push(node.val)
                node = node.right
                pre.right = null
            } else {
                pre.right = node
                node = node.left
            }
        } else {
            result.push(node.val)
            node = node.right
        }
    }
    return result
};

#class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
#    def inorderTraversal(self, root):
#        stack, result, node = [], [], root
#        while stack or node:
#            if node:
#                stack.append(node)
#                node = node.left
#            else:
#                node = stack.pop()
#                result.append(node.val)
#                node = node.right
#        return result
